File: tests/bindings/java/foo-tests/src/test/java/io/stepfunc/foo_test/IteratorTest.java
package io.stepfunc.foo_test;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.joou.Unsigned.*;

import io.stepfunc.foo.*;

class IteratorTest {

    static class TestValuesReceiver implements ValuesReceiver {
        List<org.joou.UByte> values = new ArrayList<>();

        public void onCharacters(java.util.List<StringIteratorItem> values){
            for(StringIteratorItem item : values) {
                this.values.add(item.value);
            }
        }
    }

    static class TestChunkReceiver implements ChunkReceiver {
        List<String> values = new ArrayList<>();

        public void onChunk(java.util.List<Chunk> values) {
            for(Chunk c : values) {
                java.io.ByteArrayOutputStream chars = new java.io.ByteArrayOutputStream();
                for(ByteValue bv : c.iter) {
                    chars.write(bv.value.intValue());
                }
                try {
                   this.values.add(new String(chars.toByteArray(), "UTF-8"));
                }
                catch(Exception ex) {
                   System.out.println(ex);
                   System.exit(-1);
                }

            }
        }
    }

    @Test
    void ChunkIteratorTest() {
        TestChunkReceiver receiver = new TestChunkReceiver();
        DoubleIteratorTestHelper.iterateStringByChunks("Hello World!", uint(3), receiver);
        assertThat(receiver.values.size()).isEqualTo(4);
        assertThat(receiver.values.get(0)).isEqualTo("Hel");
        assertThat(receiver.values.get(1)).isEqualTo("lo ");
        assertThat(receiver.values.get(2)).isEqualTo("Wor");
        assertThat(receiver.values.get(3)).isEqualTo("ld!");
    }

    @Test
    void StringIteratorTest() {
        TestValuesReceiver receiver = new TestValuesReceiver();
        IteratorTestHelper.invokeCallback("ABCDE", receiver);
        assertThat(receiver.values.stream()).containsExactly(ubyte(65), ubyte(66), ubyte(67), ubyte(68), ubyte(69));
    }
}

File: tests/bindings/dotnet/foo.Tests/IteratorTest.cs
using System;
using Xunit;
using foo;
using System.Linq;
using System.Collections.Generic;

namespace foo.Tests
{
    public class IteratorTest
    {
        [Fact]
        public void StringIteratorTest()
        {
            List<Byte> values = new List<Byte>();

            IteratorTestHelper.InvokeCallback("ABCDE", strings => { 
                foreach(StringIteratorItem item in strings) {
                    values.Add(item.Value);
                }
            });
            Assert.Equal(new byte[] { 65, 66, 67, 68, 69 }, values);
        }

        [Fact]
        public void ChunkIteratorTest()
        {
            List<List<Char>> values = new List<List<Char>>();
            DoubleIteratorTestHelper.IterateStringByChunks("Hello World!", 3, chunks => {
                foreach (Chunk c in chunks)
                {
                    List<Char> bytes = new List<Char>();
                    foreach (ByteValue bv in c.Iter)
                    {
                        bytes.Add(Convert.ToChar(bv.Value));
                    }
                    values.Add(bytes);
                }
            });
            Assert.Equal(4, values.Count);
            Assert.Equal(values[0], new char[] { 'H', 'e', 'l' });
            Assert.Equal(values[1], new char[] { 'l', 'o', ' ' });
            Assert.Equal(values[2], new char[] { 'W', 'o', 'r' });
            Assert.Equal(values[3], new char[] { 'l', 'd', '!' });
        }
    }
}
